packages/hasher/src/yarnWorkspaces.ts

```diff
@@ -0,0 +1,119 @@
+import * as path from "path";
+import * as fg from "fast-glob";
+
+import * as findWorkspaceRoot from "find-yarn-workspace-root";
+
+type PackageJsonWorkspaces = {
+  workspaces?:
+    | {
+        packages?: string[];
+        nohoist?: string[];
+      }
+    | string[];
+};
+
+export type WorkspaceInfo = { name: string; path: string }[];
+
+function getYarnWorkspaceRoot(cwd: string) {
+  const yarnWorkspacesRoot = findWorkspaceRoot(cwd);
+
+  if (!yarnWorkspacesRoot) {
+    throw new Error("Could not find yarn workspaces root");
+  }
+
+  return yarnWorkspacesRoot;
+}
+
+function getRootPackageJson(yarnWorkspacesRoot: string) {
+  const packageJson = require(path.join(yarnWorkspacesRoot, "package.json"));
+
+  if (!packageJson) {
+    throw new Error("Could not load package.json from workspaces root");
+  }
+
+  return packageJson;
+}
+
+function getPackages(packageJson: PackageJsonWorkspaces) {
+  const { workspaces } = packageJson;
+
+  if (workspaces && Array.isArray(workspaces)) {
+    return workspaces;
+  }
+
+  if (!workspaces || !workspaces.packages) {
+    throw new Error("Could not find a workspaces object in package.json");
+  }
+
+  return workspaces.packages;
+}
+
+function getPackagePaths(yarnWorkspacesRoot: string, packages: string[]) {
+  const packagePaths = packages.map(name =>
+    fg.sync(name, {
+      cwd: yarnWorkspacesRoot,
+      onlyDirectories: true,
+      absolute: true
+    })
+  );
+
+  return packagePaths.reduce((acc, cur) => {
+    return [...acc, ...cur];
+  });
+}
+
+export function getWorkspacePackageInfo(
+  workspacePaths?: string[]
+): WorkspaceInfo {
+  if (!workspacePaths) {
+    return [];
+  }
+
+  return workspacePaths
+    .map(workspacePath => {
+      let name: string;
+
+      try {
+        name = require(path.join(workspacePath, "package.json")).name;
+      } catch {
+        return;
+      }
+
+      return {
+        name,
+        path: workspacePath
+      };
+    })
+    .filter(Boolean) as WorkspaceInfo;
+}
+
+export function listOfWorkspacePackageNames(
```

Comment:
Do we need to export that many functions in this module? The interface of a module would be cleaner if it exposed only one concept expressed with one API (one class/one function). So here maybe we can only expose the last function?

Reply:
This is a helper function related to yarn workspaces. I think it's ok that these helper functions live here. Other functions are being exported so I can test them.